### ISSUE TITLE
localForage 1.2.2 + webpack 1.5.3 doesn't want to collaborate

### ISSUE THREAD

**Issue by ALF-er** [mozilla/localForage#344]
Hello. It seems that this PR don't fix problem of using localForage with webpack. I've created gist to illustrate problem https://gist.github.com/ALF-er/83325c24019de4edc851 (just place file index.html in ./src/ dir and main.js in ./src/js/ dir).
If I run "npm run dev" (it just copy index.html in ./dist/ dir, compile main.js with webpack and then run webpack-dev-server for ./dist/) I get:
```
WARNING in ./~/localforage/src/localforage.js
Critical dependencies:
332:20-336:22 the request of a dependency is an expression
 @ ./~/localforage/src/localforage.js 332:20-336:22

ERROR in ./~/localforage/src/drivers/localstorage.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in C:\dev\projects\wp_lf_test\node_modules\localforage\src\drivers
 @ ./~/localforage/src/drivers/localstorage.js 70:16-59

ERROR in ./~/localforage/src/drivers/websql.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in C:\dev\projects\wp_lf_test\node_modules\localforage\src\drivers
 @ ./~/localforage/src/drivers/websql.js 63:16-59
```
If in main.js uncomment second line (and comment first ofc) then it builds w/o errors but in browser console I get `localForage.getItem` - undefined function...

Maybe you can suggest something?

**Comment by tofumatt** [comment]
I don't really know much at all about webpack, any ideas @k88hudson?

-tofumatt

**Comment by gaearon** [comment]
cc @sokra

The problem seems to be due to Webpack not being able to tell if `localStorage.js` is a CommonJS or AMD file as it tries to be both.

Looks similar to https://github.com/webpack/webpack/issues/424.

**Comment by jhnns** [comment]
The problem is afaict that `localforageSerializer` is not known as AMD module when it is required. I'm not experienced with AMD modules, but can't you just use relative paths?

You could also just switch the `if` clause and check for `ModuleType.EXPORT` first . Webpack supports both module syntaxes and uses the first it encounters.

**Comment by tcoopman** [comment]
At the moment I use `require("imports?require=>false!localforage")` as suggested in https://github.com/webpack/webpack/issues/424. But it would be nice if this could be solved.

**Comment by sairion** [comment]
Having same problem. Also, using imports-loader does not solves problem. (Since it assigns var `require` as `false`, https://github.com/mozilla/localForage/blob/master/src/localforage.js#L6 throws exception. (I think Webpack could not statically analyze `require()` call inside expression)

I would like to fix this problem as well, but I ended up using IDB polyfill.

**Comment by HenrikJoreteg** [comment]
This seems like a duplicate of #290. Anyway, I'm having the same problem and have been unable to work around this. Any help appreciated.

**Comment by janza** [comment]
I've used: 

```
module: {
        loaders: [{
            test: /localforage\/dist\/localforage.js/,
            loader: 'exports?localforage',
        }]
}, 
alias: { 'localforage': 'localforage/dist/localforage.js'}
```

seems to work ok

**Comment by thgreasi** [comment]
Nice. We should either add a test for it or add it to README/wiki.

**Comment by HenrikJoreteg** [comment]
I'm still struggling with this, the webpack config tweaks suggested above don't seem to help. 

I'm getting this error:

```
WARNING in ./~/localforage/src/localforage.js
Critical dependencies:
333:24-54 the request of a dependency is an expression
 @ ./~/localforage/src/localforage.js 333:24-54

ERROR in ./~/localforage/src/drivers/localstorage.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in /Users/henrik/Sites/foo.bio/node_modules/localforage/src/drivers
 @ ./~/localforage/src/drivers/localstorage.js 70:16-59

ERROR in ./~/localforage/src/drivers/websql.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in /Users/henrik/Sites/foo.bio/node_modules/localforage/src/drivers
 @ ./~/localforage/src/drivers/websql.js 63:16-59
```

**Comment by janza** [comment]
You're right, I missed this part of the config: 
```
        noParse: [
            /localforage\/dist\/localforage.js/
        ]

```

so the whole thing is:

```
module.exports = {
    module: {
        loaders: [{
            test: /localforage\/dist\/localforage.js/,
            loader: 'exports?localforage',
        }],
        noParse: [
            /localforage\/dist\/localforage.js/
        ]
    },
    resolve: {
        alias: {
            'localforage': 'localforage/dist/localforage.js'
        }
    }
};


```

**Comment by prakhar1989** [comment]
@janza: Hi, I've modified my `webpack.config.js` as suggested by you and I can no longer see the errors that @HenrikJoreteg was seeing. However, now I'm getting the following -

```
ERROR in ./js/stores/SurveyStore.js
Module not found: Error: Cannot resolve module 'exports' in /Users/asrinivasan/Code/react-surveyman/js/stores
 @ ./js/stores/SurveyStore.js 13:0-22
```

The `SurveyStore` looks like this - 
```javascript
var localForage = require('localforage');
```

Any help?

**Comment by janza** [comment]
you need to install exports-loader: ```npm i -S exports-loader```

**Comment by prakhar1989** [comment]
Thanks @janza. Although I'm now using another `localStorage` [lib](https://github.com/tsironis/lockr) I'll give this a shot and report if there are any issues. Thanks again!

**Comment by shprink** [comment]
Thanks @janza it works as expected :)

**Comment by thgreasi** [comment]
Any chance you ca check if #415 fixes this?

**Comment by ALF-er** [comment]
webpack 1.11.0, localforage 1.2.7
Still get
```
WARNING in ./~/localforage/src/localforage.js
Critical dependencies:
334:24-54 the request of a dependency is an expression
 @ ./~/localforage/src/localforage.js 334:24-54

ERROR in ./~/localforage/src/drivers/localstorage.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in C:\dev\projects\trill\node_modules\localforage\src\drivers
 @ ./~/localforage/src/drivers/localstorage.js 70:16-59

ERROR in ./~/localforage/src/drivers/websql.js
Module not found: Error: Cannot resolve module 'localforageSerializer' in C:\dev\projects\trill\node_modules\localforage\src\drivers
 @ ./~/localforage/src/drivers/websql.js 63:16-59
```
for me =(

**Comment by tofumatt** [comment]
Can you try with the latest master instead? 1.2.7 wouldn’t have this fix.


- tofumatt

**Comment by thgreasi** [comment]
@ALF-er OOOOoooops!, my bad.
The actual webpack changes are in a commit that I didn't push.
Now that the PR is merged, I will do a rebase as soon as I'm home. Sorry for the inconvenience/false positive.

**Comment by tofumatt** [comment]
Oops, haha! 😂


- tofumatt

**Comment by thgreasi** [comment]
Just opened #417 that directs webpack to use the dist file of the repo.
This should remove the need of aliasing the module in your webpack configuration.
Tested with webpack v1.11 on the how-to-repo [MattKunze/localforage-webpack](https://github.com/MattKunze/localforage-webpack).
Any feedback will be much appreciated.

**Comment by ALF-er** [comment]
So, after `npm install git+https://git@github.com/mozilla/localForage.git`
and setup webpack to run localForage.js thru babel-loader I got this:

```
ERROR in ./~/localforage/src/localforage.js
Module build failed: ReferenceError: Unknown plugin "system-import-transformer"
    at PluginManager.subnormaliseString (C:\dev\projects\trill\node_modules\babel-core\lib\transformation\file\plugin-manager.js:147:13)
    at PluginManager.add (C:\dev\projects\trill\node_modules\babel-core\lib\transformation\file\plugin-manager.js:190:40)
    at File.buildTransformers (C:\dev\projects\trill\node_modules\babel-core\lib\transformation\file\index.js:237:21)
    at new File (C:\dev\projects\trill\node_modules\babel-core\lib\transformation\file\index.js:139:10)
    at Pipeline.transform (C:\dev\projects\trill\node_modules\babel-core\lib\transformation\pipeline.js:164:16)
    at transpile (C:\dev\projects\trill\node_modules\babel-loader\index.js:12:22)
    at Object.module.exports (C:\dev\projects\trill\node_modules\babel-loader\index.js:69:12)
 @ ./src/app/utils/OfflineStorageUtils.js 11:19-41
```

**Comment by thgreasi** [comment]
Can try using the branch in #417? Hopefully, you shouldn't have to do any configuring or use babel at all, since package.json should direct webpack to use the dist files.

**Comment by HenrikJoreteg** [comment]
@thgreasi i just tried setting "webpack" to dist/localforage.js as per your fix. I'm still seeing issues:

```
WARNING in ./~/localforage/dist/localforage.js
Critical dependencies:
2603:24-54 the request of a dependency is an expression
 @ ./~/localforage/dist/localforage.js 2603:24-54

WARNING in ./~/localforage/dist/localforage.nopromises.js
Critical dependencies:
1920:24-54 the request of a dependency is an expression
 @ ./~/localforage/dist/localforage.nopromises.js 1920:24-54

WARNING in ./~/localforage/dist/localforage.min.js
Critical dependencies:
7:24233-24247 the request of a dependency is an expression
 @ ./~/localforage/dist/localforage.min.js 7:24233-24247

WARNING in ./~/localforage/dist/localforage.nopromises.min.js
Critical dependencies:
7:19049-19063 the request of a dependency is an expression
 @ ./~/localforage/dist/localforage.nopromises.min.js 7:19049-19063

ERROR in ./~/localforage/dist/localforage.js
Module not found: Error: Cannot resolve 'file' or 'directory' ./../utils/serializer in /Users/henrik/Sites/hubflow-redux/node_modules/localforage/dist
 @ ./~/localforage/dist/localforage.js 1596:24-56 1918:24-56

```

It keeps going for a bit..

**Comment by thgreasi** [comment]
[Here is my test branch of @MattKunze's localforage-webpack.](https://github.com/thgreasi/localforage-webpack)
It uses [the branch of](https://github.com/thgreasi/localforage/tree/webpack-fix) PR #417.
I'm just running webpack in terminal and the index.html page seems to be working.

I think I'm not doing using magic webpack configuration for localforage to work.
Am I doing something wrong?

Any chance I can reach some of the configuration of your app?

**Comment by thgreasi** [comment]
@HenrikJoreteg all I get in terminal is:
```
~/Desktop/localforage-webpack$ webpack
Hash: 34b97eaf37c1224e529b
Version: webpack 1.11.0
Time: 105ms
    Asset    Size  Chunks             Chunk Names
bundle.js  102 kB       0  [emitted]  main
   [0] ./index.js 148 bytes {0} [built]
    + 3 hidden modules
```

**Comment by thgreasi** [comment]
Added grunt to the test repo and realized that it was already using [webpack/script-loader](https://github.com/webpack/script-loader). I hope that importing localforage with something like: `require('script!localforage');` isn't too bad at the moment.

**Comment by thgreasi** [comment]
Just pushed an extra commit to PR #417, that adds webpack bundling tests.

**Comment by luisrudge** [comment]
so, can anyone please test and merge this if it's really working? :D thaaanks

**Comment by thgreasi** [comment]
#417 has been merged into master, but I think we have yet to release it under a new tag.
v1.2.7 should work for now.
We are probably going for a proper fix in v1.3.

**Comment by tofumatt** [comment]
Looks like this should now be fixed in 1.2.10. #439 fixed it up, if there's an issue let me know (be sure to following the README's guide to using it in 1.2.10: https://github.com/mozilla/localForage/tree/1.2.10#browserify-and-webpack
